### ISSUE TITLE
Packer, ansible-remote and local_action

### ISSUE THREAD

**Issue by ncrothe** [hashicorp/packer#3925]
Hi,

I'm relatively new to packer and ansible and I ran into an issue using packer's (v0.10.1) ansible-remote module with the docker builder and a playbook containing a local_action.

My understanding is, that ansible-remote runs on the host (i.e. the machine running packer) and executes tasks on the remote box, in this case a docker instance, via ssh pipe(?). That means a local_action should **not** be run on the remote box.

I have tried the simplest possible repo at
https://gist.github.com/ncrothe/08c368313570e409be9f8481063f22eb

This runs hostname and prints out via debug the expected outcome would be the host's hostname, the observed outcome is the docker instance's hostname.

Am I completely misunderstanding how this is supposed to work or is this actually broken?

Thanks!

**Comment by rickard-von-essen** [comment]
Please [gist](https://gist.github.com/) a full debug log from `PACKER_LOG=1 packer build template.json`.

**Comment by ncrothe** [comment]
Ah, thought I did but was obviously wrong. I've updated the gist. https://gist.github.com/ncrothe/08c368313570e409be9f8481063f22eb

**Comment by bhcleek** [comment]
@ncrothe I believe the problem here is that you're bypassing the Ansible provisioner and causing Ansible to communicate directly with your DOCKER_HOST by setting `ansible_connection=docker`.

Unfortunately, Packer 0.10.1's Docker builder doesn't play nicely with the Ansible provisioner. Can you remove the `ansible_connection=docker` value from `extra-vars` and try again with Packer 0.11?

**Comment by mwhooker** [comment]
Going to close this for now. If this is still an issue after the above fix, please open a new issue.

**Comment by mchagan** [comment]
I tried removing the `ansible_connection={{user `ansible_connection`}}` value from `extra-vars` in local_action.json and received an error.  Is this what you meant?  I am running Packer 0.12.2.

**Comment by bhcleek** [comment]
Yes, that's what I meant. You'll need to remove the `ansible_connection` and the `ansible_host` values from `extra-vars`.

**Comment by mchagan** [comment]
I tried removing the --extra-vars and I am getting errors.  Could you please look at the information in gist: https://gist.github.com/mchagan/cd6fe5f8a12853c54e754b6b5c447112 and tell me what I am doing wrong?

**Comment by bhcleek** [comment]
I'm not certain, but I think it may have to do with having `gather_facts` implicitly enabled.

There is a test that exercises using the ansible-remote provisioner with the docker builder. Here are the relevant files from that test:
https://github.com/mitchellh/packer/blob/master/test/fixtures/provisioner-ansible/docker.json
https://github.com/mitchellh/packer/blob/master/test/fixtures/provisioner-ansible/playbook.yml

**Comment by mchagan** [comment]
I was able to reproduce this.  Removing "--extra-vars", "ansible_host={{user `ansible_host`}} ansible_connection={{user `ansible_connection`}}" did the trick.  

stdout changed from "2c746a0e3caa" to "OLC-1265"

Before: https://gist.github.com/mchagan/c3a470cf218670fc384c8347cbfa4e70
After: https://gist.github.com/mchagan/e95dfcba36a5d52fbc985d0ff9062c1d

I would like some verification before I attempt to fix it.

**Comment by bhcleek** [comment]
@mchagan. What would you expect to happen? The playbook is using a local action to get the hostname; Ansible is literally running on your host; `OLC-1265` is the correct output.

**Comment by mchagan** [comment]
Just making sure this is a bug.  The output should be "OLC-1265" with "--extra-vars" set to "ansible_host={{user ansible_host}} ansible_connection={{user ansible_connection}}".
Thank you for the verification.

**Comment by bhcleek** [comment]
It doesn't sound like a bug to me. What is it that you believe is incorrect about what you're seeing?

**Comment by mchagan** [comment]
This is my understanding. ncrothe reported a bug that the output was the wrong hostname.  

Running the original code supplied by ncrothe on my host OLC-1265, the output from hostname is "2c746a0e3caa".  You suggested on Oct 30, 2016 to remove the "ansible_host={{user ansible_host}} and ansible_connection={{user ansible_connection}}" from "--extra-vars".  
I did so and the hostname output is "OLC-1265" which is the machine I am running packer from.  

I may be mistaken but I thought the output from the hostname should be "OLC-1265" whether or not ansible_host={{user ansible_host}} and ansible_connection={{user ansible_connection}}" is specified in "--extra-vars".  I apologize if this is not a bug and wasting your time.

**Comment by bhcleek** [comment]
I'm happy to help clarify things; no apologies necessary.

The behavior originally reported was due to the use of the docker connection type for ansible. If the playbook outputs the wrong thing when using a local_action with the docker connection type, then the problem lies with ansible, not packer's ansible provisioner.

**Comment by mchagan** [comment]
Ah, sorry about that.  I'll look at ansible.  Thanks for the explanation.

**Comment by mchagan** [comment]
I finally got back to this and did some more research and tests.  @bhcleek, you were right all along.  It's not a bug with packer or ansible (IMHO).  The argument "ansible_host" in "--extra-vars" executes the hostname command on the host defined by "ansible_host" that is why the host name is not with you expected.  I think this can be closed.  I apologize if I wasted anyone's time.

**Comment by mwhooker** [comment]
glad you got it working

**Comment by bhcleek** [comment]
Thank you for the confirmation.
